### PR TITLE
Create unit tests #45

### LINKED ISSUE #45: Create unit tests
We need unit tests!

## Specs
- Preferably use Pythons built-in `unittest` framework.
- The tests should be in a script called: `test_ask.py`
- The most important methods to test are (in no particular order):
  - `get_root_from_file_path()`
  - `route_path_to_func_name()`
  - `maybe_place_space_before()`
  - `route_params()`
  - `get_current_tab_level()`
  - `parser()` ❗️
  - `lex_var_keyword()`
  - `add_part()`
  - `fix_up_code_line()`
  - `lexer()` ❗️
 
(❗️= complicated tests required)

### PR DESCRIPTION
**Pull request by pillemer**
Added first new test for maybe_place_space_before()

Not sure how often one should commit/push/create pull requests so please let me know what you prefer.

Closes #45.

### CODE REVIEW

**Comment by pillemer** [comment]
Added tests for route_params().
It seems the online opinion is that I should write a specific test per assertion so I tried it out this time. Please let me know which method you prefer.
You'll notice the test fails on the last three assertions. Please let me know if my assumptions for the expected results there were wrong.

**Review by Buscedv — COMMENTED** [review]
Awesome work 👍  ! Thank you for your investment in this project, it means a lot.

- Your assumptions in the last three tests are correct.

This means that bugfixes in `query_params()` are needed. I'll update Buscedv/Ask's *ask.py* file with fixes today. I'll let you know when it's done. You can then either create a pull request into your fork from Buscedv/Ask (master) to pillemer/Ask (master), or you could straight up copy-paste ask.py into your fork's ask.py file (whichever you prefer).

- I think that you are also right about having multiple tests instead of putting them in a loop. So, you can update `test_maybe_place_space_before()` and also `test_get_root_from_file_path()`.

Thanks again!

**Comment by Buscedv** [comment]
I have now updated *ask.py* in Buscedv/Ask (master) with the bugfixes in `query_params()`, the last three tests should now all pass.

**Comment by Buscedv** [comment]
> Added first new test for maybe_place_space_before()
> 
> Not sure how often one should commit/push/create pull requests so please let me know what you prefer.
> 
> Closes #45.

Oh and regarding your question about how often you should commit: It's really up to you but basically commit and push whenever you start working on something new/different. Basically whenever a natural separation of tasks happen.

File: test_ask.py [hunk 1]
```diff
@@ -24,3 +24,78 @@ def test_route_path_to_func_name(self):
...
+			'string % ',
+			'string.% ',
+		]
+
+		for index, value in enumerate(parsed):
+			self.assertEqual(expected_result[index], ask.maybe_place_space_before(value, token_val))
+			
+	def test_route_params_empty_string(self):
+		test_value = ''
+		actual = ask.route_params(test_value)
```
[the remark this answers is not in the document]
Comment (pillemer) [reply test_ask.py:56]: I've removed the unnecessary line.

**Comment by pillemer** [comment]
I've broken the other two functions into multiple tests.
A couple are failing but again, it could be my assumption for the expected result that is wrong.

Thank you for the detailed responses btw, they are super helpful! You're really making this rather intimidating concept of open source contributing much easier to do for a novice such as myself.

File: test_ask.py [hunk 2]
```diff
@@ -3,24 +3,131 @@
...
+		expected = 'string % '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string.' 
+		expected = 'string.% '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	#route_params
```
Comment (Buscedv) [inline test_ask.py:85]: Put a space after #

File: test_ask.py [hunk 3]
```diff
@@ -3,24 +3,131 @@
...
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_param_in_param(self):
+		test_value = '<par<param>ram>'
+		expected = 'param'
+		self.assertEqual(expected, ask.route_params(test_value))
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (Buscedv) [inline test_ask.py:133]: There was an indentation warning here when I opened this file in my text editor. Make sure that this line uses tabs and not spaces.

File: test_ask.py [hunk 4]
```diff
@@ -3,24 +3,131 @@
...
+		self.assertEqual(expected, ask.route_path_to_func_name(test_value))
+
+	def test_route_path_to_func_name_path_char(self):
+		test_value = '/path/<path>'
+		expected = '_path_path'
+		self.assertEqual(expected, ask.route_path_to_func_name(test_value))
+
+	def test_route_path_to_func_name_path_slash_dash_path(self):
+		test_value = '/path/-path'
+		expected = '_path_path'
```
Comment (Buscedv) [inline test_ask.py:39]: Expected should be: `_path__path` the dash after the slash also gets converted into an underscore.

File: test_ask.py [hunk 5]
```diff
@@ -3,24 +3,131 @@
...
+		self.assertEqual(expected, ask.route_path_to_func_name(test_value))
+
+	def test_route_path_to_func_name_path_dash_path(self):
+		test_value = '/path-path'
+		expected = '_path_path'
+		self.assertEqual(expected, ask.route_path_to_func_name(test_value))
+
+	def test_route_path_to_func_name_path_char(self):
+		test_value = '/path/<path>'
+		expected = '_path_path'
```
Comment (Buscedv) [inline test_ask.py:34]: Expected should be: `_path__path_` since *<* and *>*s also gets converted into underscores.

**Review by Buscedv — CHANGES_REQUESTED** [review]
Keep up the good work 🚀!

The two failing tests should be fixed if you update to `expected` variables according to my review comments.

I took a look at the `route_path_to_func_name()` function in *ask.py* and noticed that it was written in a very strange way (don't know what I was thinking at the time 🤦‍♂️). There is now an updated version in `Buscedv/Ask`'s master branch.

File: test_ask.py (hunk 2 above)
Comment (pillemer) [reply test_ask.py]: Done.

File: test_ask.py (hunk 3 above)
Comment (pillemer) [reply test_ask.py]: I always use tabs so I don't know what triggered it. I've written it fresh just in case.

File: test_ask.py (hunk 4 above)
Comment (pillemer) [reply test_ask.py]: No problem. I've updated my expected results.

File: test_ask.py (hunk 5 above)
Comment (pillemer) [reply test_ask.py]: Updated the expected results to reflect this.

**Comment by pillemer** [comment]
Just to make things a little easier to track.
So far we have tests for the following functions:
- [x] get_root_from_file_path()
- [x] route_path_to_func_name()
- [x] maybe_place_space_before()
- [x] route_params()
- [x] get_current_tab_level()
- [ ] parser() ❗️
- [ ] lex_var_keyword()
- [x] add_part()
- [ ] fix_up_code_line()
- [ ] lexer() ❗️
- [x] insert_indention_group_markers
- [x] tokens_grouped_by_lines 

@Buscedv, what do you mean by (❗️ =complicated tests required)? 

Additional functions that could use testing?
- [ ] transpile_var
- [ ] transpile_keyword
- [ ] transpile_decorator
- [ ] transpile_db_action

**Comment by Buscedv** [comment]
@pillemer "complicated tests required" means that the function requires quite long tests, ex. parse is really hard to test properly without making it parse an entire app, same goes for lex. They work in such a way that it is no point to test them with simple one-line strings, those sorts of tests will probably always pass. The tests that are needed to really make sure that these functions work as expected is to make them lex/parse entire small applications/programs.

```
 transpile_var()
 transpile_keyword()
 transpile_decorator()
 transpile_db_action()
```

You can write tests for these if you want but many of them are simply dictionary lookups, so they are low on the list for functions that need testing. But fell free to make tests for them anyway, it doesn't hurt.

File: test_ask.py [hunk 6]
```diff
@@ -82,52 +82,236 @@ def test_maybe_place_space_before(self):
...
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['VAR', 'variable'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FORMAT', '\t'],
+			['VAR', 'variable'],
+		]
+
+		excpected = [
```
Comment (Buscedv) [inline test_ask.py]: whoops, I misspelt "expected" here, my bad...

File: test_ask.py [hunk 7]
```diff
@@ -82,52 +82,236 @@ def test_maybe_place_space_before(self):
...
 
+	def test_add_part_string_false_code_without_newline(self):
+		is_string = False
+		code = 'code'
+		expected = (
+			[{'is_string': False, 'code': 'code'}], 
+			'', 
+			True
+		)
+		self.assertEqual(expected, ask.add_part([],is_string, code))
```
Comment (Buscedv) [inline test_ask.py:314]: space after `,`

File: test_ask.py [hunk 8]
```diff
@@ -82,52 +82,236 @@ def test_maybe_place_space_before(self):
...
 
 if __name__ == '__main__':
```
Comment (Buscedv) [inline test_ask.py:316]: Add one more empty line before `if __name___ ...`

File: test_ask.py [hunk 9]
```diff
@@ -82,52 +82,236 @@ def test_maybe_place_space_before(self):
...
+		]
+
+		self.assertEqual(expected, ask.insert_indention_group_markers(tokens))
+
+	# add_part
+	def test_add_part_string_true_code_with_newline(self):
+		is_string = True
+		code = 'code \n'
+		expected = (
+			[{'is_string': True, 'code': 'code \n'}],
```
Comment (Buscedv) [inline test_ask.py:280]: `add_part()` will always return `code` the same way as it gets passed in, so no need to hardcode `'code': 'code \n'` when you have already defined `code` on line 278. Do this instead: `'code': code` Do this in all of `add_part()`'s tests.

**Review by Buscedv — CHANGES_REQUESTED** [review]
Fantastic! 👍

File: test_ask.py (hunk 6 above)
Comment (pillemer) [reply test_ask.py]: fixed

File: test_ask.py (hunk 7 above)
Comment (pillemer) [reply test_ask.py:314]: removed

File: test_ask.py (hunk 8 above)
Comment (pillemer) [reply test_ask.py:316]: done

File: test_ask.py (hunk 9 above)
Comment (pillemer) [reply test_ask.py:280]: done.

**Comment by pillemer** [comment]
I've added tests for the four 'transpile' functions as well as for lex_var_keyword. 
As for the remaining three functions:
- [ ] parser() 
- [ ]  fix_up_code_line()
- [ ]  lexer() 

I'm not sure I know how to test these ones.

Also, I noticed in lex_var_keyword() that you declare a global variable 'variables' that, unless I'm mistaken, doesn't get used in the function.

**Comment by Buscedv** [comment]
> I've added tests for the four 'transpile' functions as well as for lex_var_keyword.
> As for the remaining three functions:
> 
> * [ ]  parser()
> * [ ]   fix_up_code_line()
> * [ ]   lexer()
> 
> I'm not sure I know how to test these ones.
> 
> Also, I noticed in lex_var_keyword() that you declare a global variable 'variables' that, unless I'm mistaken, doesn't get used in the function.
> 
> Should I mark this PR as ready to review?

I can write tests for the three remaining functions. You've done an excellent job with these tests!
- I haven't noticed the unnecessary import of `variables` before, you can delete that line.
- You can mark this PR as ready.

**Comment by pillemer** [comment]
Thank you for giving me the opportunity! I'd never made tests like this before and working to try and figure out what somebody else's code is actually trying to achieve was very fun and challenging. I'm glad I was able to help.

**Review by Buscedv — APPROVED** [review]
Thank you so much for your help! It's been a pleasure working with you.

**Pull request merged**